### PR TITLE
TakenName rule

### PR DESCRIPTION
**Pull request by msoedov**
This PR adds TakenName rule described in #151 

Todos: 
--------
- [x] Extend keywords list
- [x] Clean up

### CODE REVIEW

File: lib/dogma/rule/taken_name.ex [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      {node, [error( meta[:line], name ) | errors]}
+    else
+      {node, errors}
+    end
+  end
+
+  defp error(pos, name) do
+    %Error{
+      rule:     __MODULE__,
+      message:  "`#{name}` is already taken and overrides standart library",
```
Comment (lpil) [inline lib/dogma/rule/taken_name.ex]: Typo: Standard

:)

File: test/dogma/rule/taken_name_test.exs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      end
+    end
+    """ |> lint
+    assert [error_on_line(1, :require)] == errors
+  end
+
+  defp error_on_line(line, name) do
+    %Error{
+      line: Dogma.Script.line(line),
+      message: "`#{name}` is already taken and overrides standart library",
```
Comment (lpil) [inline test/dogma/rule/taken_name_test.exs]: Typo here too

File: lib/dogma/rule/taken_name.ex [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  @behaviour Dogma.Rule
+  alias Dogma.Script
+  alias Dogma.Error
+
+  reserved_words = ~w(alias bc case cond exit function if import inbits inlist
+               is_atom is_binary is_bitstring is_boolean is_exception is_float
+               is_function is_integer is_list is_number is_pid is_port
+               is_record is_reference is_tuple lc quote raise receive require
+               respawn super throw try unless unless unquote use
+               )
```
Comment (msoedov) [inline lib/dogma/rule/taken_name.ex]: Do you have any ideas what else we can add to this list?

Comment (lpil) [inline lib/dogma/rule/taken_name.ex]: Perhaps we could grab all the function and macro names from Kernel and SpecialForms?

File: lib/dogma/rule/taken_name.ex [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    test_predicate(name, meta, node, errors)
+  end
+  defp check_node({:defp, _, [{name, meta, _}|_]} = node, errors) do
+    test_predicate(name, meta, node, errors)
+  end
+  defp check_node(node, errors) do
+    {node, errors}
+  end
+
+  defp test_predicate(name) do
```
Comment (lpil) [inline lib/dogma/rule/taken_name.ex]: Can we give this function a more descriptive name?

File: lib/dogma/rule/taken_name.ex [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    if HashSet.member?(@keywords, name) do
+      name
+    else
+      nil
+    end
+  end
+
+  defp test_predicate({:unquote,_,_} , _meta, node, errors) do
```
Comment (lpil) [inline lib/dogma/rule/taken_name.ex]: It feels like this function and the previous one should be two different functions. They have completely different type signatures, and perform different roles. Both could do with clearer names.

File: test/dogma/rule/taken_name_test.exs [hunk 6]
```diff
@@ -0,0 +1,69 @@
+defmodule Dogma.Rule.TakenNameTest do
+  use RuleCase, for: TakenName
+  use ShouldI
+
+  defp lint(script) do
+    script |> Script.parse!( "" ) |> fn s -> TakenName.test(@rule, s) end.()
```
Comment (lpil) [inline test/dogma/rule/taken_name_test.exs]: Rules should be dispatched to using the protocol, not by invoking the module function directly.

File: test/dogma/rule/taken_name_test.exs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  end
+
+  defp verify fn_name do
+    errors = """
+    defp #{fn_name} do
+      :function_body
+    end
+    """ |> lint
+    assert [error_on_line(1, String.to_atom(fn_name))] == errors,
```
Comment (lpil) [inline test/dogma/rule/taken_name_test.exs]: Don't cast to atoms when you can avoid it.

File: lib/dogma/rule/taken_name.ex (hunk 5 above)
Comment (msoedov) [inline lib/dogma/rule/taken_name.ex]: I'm not sure that we need this so I going to remove the condition

**Comment by lpil** [comment]
Great, thanks!

**Comment by msoedov** [comment]
Happy to contribute!

**Pull request merged**